zebra-consensus/src/block/check.rs

```diff
@@ -143,17 +149,37 @@ pub fn subsidy_is_valid(block: &Block, network: Network) -> Result<(), BlockErro
             .collect();
         let output_amounts = subsidy::general::output_amounts(coinbase);
 
+        // funding stream addresses
+        let mut found_outputs = HashSet::<FundingStreamReceiver>::new();
+        for receiver in FundingStreamReceiver::receivers() {
+            let address =
+                subsidy::funding_streams::funding_stream_address(height, network, receiver);
+
+            let amount = *funding_streams
+                .get(&receiver)
+                .expect("funding_streams hashmap has all possible receivers.");
+
+            // we should have at least one output with receiver address and amount
+            let outputs = subsidy::funding_streams::find_output_with_address_and_amount(
+                coinbase, address, amount,
+            );
+            if !outputs.is_empty() {
+                found_outputs.insert(receiver);
+            }
+        }
+
```

Comment:
I think we might need to check if the amount is correct for the receiver. If I'm not mistaken, with the current code we check if we have all receivers and all amounts, but not if the amount is mapped to the correct receiver.

I'm not sure this is correct, but one idea might be to return early after each check (note: I couldn't select the non-edited lines to include in the suggestion):

```suggestion
    } else if halving_div < 4 {
        // Funding streams are paid from Canopy activation to the second halving
        // Note: Canopy activation is at the first halving on mainnet, but not on testnet
        // ZIP-1014 only applies to mainnet, ZIP-214 contains the specific rules for testnet

        // funding stream amount values
        let funding_streams = subsidy::funding_streams::funding_stream_values(height, network)
            .expect("We always expect a funding stream hashmap response even if empty");

        // Consensus rule:[Canopy onward] The coinbase transaction at block height `height`
        // MUST contain at least one output per funding stream `fs` active at `height`,
        // that pays `fs.Value(height)` zatoshi in the prescribed way to the stream's
        // recipient address represented by `fs.AddressList[fs.AddressIndex(height)]
        for (receiver, expected_amount) in funding_streams {
            let address =
                subsidy::funding_streams::funding_stream_address(height, network, receiver);

            let output_amount: Amount<NonNegative> =
                subsidy::funding_streams::find_output_with_address(coinbase, address)
                    .map(transparent::Output::value)
                    .fold(Ok(Amount::zero()), |sum, item| sum + item)
                    .map_err(|_| SubsidyError::FundingStreamValueNotFound)?;

            if output_amount != expected_amount {
                Err(SubsidyError::FundingStreamValueNotFound)?;
            }
        }

        Ok(())
```

Reply:
I agree, the consensus rule says that we need at least one output that sends the amount to the address:
> The coinbase transaction at block height `height`
> MUST contain at least one output per funding stream `fs` active at `height`,
> that pays `fs.Value(height)` zatoshi ... to the stream's recipient address

Reply:
I don't think we're allowed to sum the amounts of multiple outputs to the same funding stream address.

Instead, at least one of the outputs must have the correct amount. But it doesn't matter which output it is, or what order it is in.

So I think `HashSet<(Address, Amount)>::subset` is the check you want here.

Reply:
Extending the idea a bit, we can avoid creating a new `HashSet` by just checking if we find the pair:

```rust
let has_expected_output =
    subsidy::funding_streams::find_output_with_address(coinbase, address)
        .map(transparent::Output::value)
        .find(|value| value == expected_value)
        .is_some();

if !has_expected_output {
    Err(SubsidyError::FundingStreamValueNotFound)?;
}
```

Reply:
Thanks, i applied the suggestions discussed here at https://github.com/ZcashFoundation/zebra/pull/3040/commits/0f43eaf2b4ad4b2090db40b34f933decd96fbf87